File: demo/healthhub/frontend/src/models/Doctor.ts
/**
 * Doctor domain model.
 */

export interface Doctor {
  readonly id: string
  readonly userId: string
  readonly firstName: string
  readonly lastName: string
  readonly specialization: string
  readonly licenseNumber: string
  readonly canPrescribe: boolean
  readonly phone: string | null
  readonly email: string
}

// API response type (matches backend)
export interface DoctorApiResponse {
  readonly id: string
  readonly user_id: string
  readonly first_name: string
  readonly last_name: string
  readonly specialization: string
  readonly license_number: string
  readonly can_prescribe: boolean
  readonly phone: string | null
  readonly email: string
}

// Converter from API response to domain model
export const fromApiResponse = (response: DoctorApiResponse): Doctor => ({
  id: response.id,
  userId: response.user_id,
  firstName: response.first_name,
  lastName: response.last_name,
  specialization: response.specialization,
  licenseNumber: response.license_number,
  canPrescribe: response.can_prescribe,
  phone: response.phone,
  email: response.email,
})

// Display helpers
export const getFullName = (doctor: Doctor): string =>
  `Dr. ${doctor.firstName} ${doctor.lastName}`

export const getDisplayTitle = (doctor: Doctor): string =>
  `${getFullName(doctor)} - ${doctor.specialization}`

File: demo/healthhub/backend/app/domain/doctor.py
"""Doctor domain model."""

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from effectful.domain.optional_value import OptionalValue


@dataclass(frozen=True)
class Doctor:
    """Doctor entity with credentials.

    Immutable domain model following Effectful patterns.

    Attributes:
        id: Unique doctor identifier
        user_id: Associated user account UUID
        first_name: Doctor's first name
        last_name: Doctor's last name
        specialization: Medical specialization
        license_number: Medical license number
        can_prescribe: Whether authorized to prescribe medications
        phone: Optional phone number
        created_at: Timestamp when doctor record was created
        updated_at: Timestamp of last update
    """

    id: UUID
    user_id: UUID
    first_name: str
    last_name: str
    specialization: str
    license_number: str
    can_prescribe: bool
    phone: OptionalValue[str]
    created_at: datetime
    updated_at: datetime
